python/cudf/cudf/core/column/column.py

```diff
@@ -1948,9 +1948,15 @@ def as_column(
                 raise TypeError(
                     f"Cannot convert a {inferred_dtype} of object type"
                 )
+            elif (
+                not nan_as_null
+                and inferred_dtype == "string"
+                and any(isinstance(x, float) and np.isnan(x) for x in arbitrary)
```

Comment:
In addition to my earlier question about whether we can optimize this more, do we also need to handle `complex` objects here, like `np.nan * 1j`? Just want to ensure that `isinstance(..., float)` is sufficient.